examples/directml/llm/user_script.py

```diff
@@ -0,0 +1,224 @@
+        )
+    else:
+        inputs["position_ids"] = torch.zeros((batch_size, sequence_length), dtype=torch.int64)
+        inputs["input_ids"] = torch.zeros((batch_size, sequence_length), dtype=torch.int64)
+
+    return inputs
+
+
+def decoder_ort_inputs(batch_size):
+    sequence_length = 10
+    max_seq_len = 1024
+    head_size = config.hidden_size // config.num_heads
+
+    inputs = {
+        "attention_mask": torch.zeros((batch_size, max_seq_len), dtype=torch.int64),
+    }
+
+    for layer_idx in range(config.num_layers):
+        inputs[f"past_key_values.{layer_idx}.key"] = torch.rand(
+            (batch_size, config.num_key_value_heads, max_seq_len, head_size), dtype=torch.float32
+        )
+        inputs[f"past_key_values.{layer_idx}.value"] = torch.rand(
+            (batch_size, config.num_key_value_heads, max_seq_len, head_size), dtype=torch.float32
+        )
+
+    if config.model_type == "llava":
+        channel_count = 3
+        image_size = 336
+        inputs["pixel_values"] = torch.zeros((batch_size, channel_count, image_size, image_size), dtype=torch.float32)
+
+        # 32000 is the value for the image token and needs to be be there for the model to be successfully generated
+        inputs["input_ids"] = torch.nn.functional.pad(
+            torch.zeros((batch_size, sequence_length - 1), dtype=torch.int64), (0, 1), value=32000
+        )
+    else:
+        inputs["position_ids"] = torch.zeros((batch_size, sequence_length), dtype=torch.int64)
+        inputs["input_ids"] = torch.zeros((batch_size, sequence_length), dtype=torch.int64)
+
+    return inputs
+
+
+def decoder_ort_data_loader(data_dir, batch_size, *args, **kwargs):
+    return RandomDataLoader(decoder_ort_inputs, batch_size)
+
+
+# -----------------------------------------------------------------------------
+# Quantization calibration
+# -----------------------------------------------------------------------------
+
+
+def tokenize_function(examples):
+    # There's a bug that makes the rust-based fast tokenizer hang randomly (probably due to a deadlock),
+    # so use the "slow" python one instead
+    tokenizer = AutoTokenizer.from_pretrained(config.model_id, use_fast=False)
+    return tokenizer(examples["text"])
+
+
+class PileDataloader:
+    def __init__(self, model_path, batch_size=1, seqlen=2048, max_seq_len=2080, sub_folder="train"):
+        random.seed(0)
+        self.seqlen = seqlen
+        self.max_seq_len = max_seq_len
+        self.batch_size = batch_size
+        self.head_size = config.hidden_size // config.num_heads
+        self.dataset = load_dataset("NeelNanda/pile-10k", split=sub_folder)
+        self.dataset = self.dataset.map(tokenize_function, batched=True)
+        self.dataset.set_format(type="torch", columns=["input_ids", "attention_mask"])
+        self.sess = None
+
+    def __iter__(self):
+        try:
+            while True:
```

Comment:
this seems an endless loop?

Reply:
It's just how generators with `yield` work. We don't know how many times they're gonna call it, so we need to have an "endless" loop that keeps yielding until `StopIteration` is raised, at which point we stop the loop.

Reply:
the problem here is that the StopIteration will never be raised.

Reply:
as https://docs.python.org/3/library/exceptions.html#StopIteration said, the StopIteration is raised when calling next function. But I don't find the next call in the method.